app/models/detector/suggested_resource.rb

```diff
@@ -0,0 +1,57 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: detector_suggested_resources
+#
+#  id          :integer          not null, primary key
+#  title       :string
+#  url         :string
+#  phrase      :string
+#  fingerprint :string
+#  created_at  :datetime         not null
+#  updated_at  :datetime         not null
+#
+
+require 'stringex/core_ext'
+
+module Detector
+  # Detector::SuggestedResource stores custom hints that we want to send to the
+  # user in response to specific strings. For example, a search for "web of
+  # science" should be met with our custom login link to Web of Science via MIT.
+  class SuggestedResource < ApplicationRecord
+    before_save :update_fingerprint
+
+    def update_fingerprint
+      self.fingerprint = calculate_fingerprint
+    end
+
+    # This implements the OpenRefine fingerprinting algorithm. See
+    # https://openrefine.org/docs/technical-reference/clustering-in-depth#fingerprint
+    def calculate_fingerprint
+      temp = phrase
+      temp.strip!
+      temp.downcase!
+
+      # This removes all punctuation and symbol characters from the string.
+      temp.gsub!(/\p{P}|\p{S}/, '')
+
+      # Normalize to ASCII (e.g. gödel and godel are liable to be intended to
+      # find the same thing)
+      temp = temp.to_ascii
+
+      # Coercion to ASCII can introduce new symbols, so we remove those now.
+      temp.gsub!(/\p{P}|\p{S}/, '')
+
+      # Tokenize
+      array = temp.split
+
+      # Remove duplicates and sort
+      array.uniq!
+      array.sort!
+
+      # Rejoin tokens
+      array.join(' ')
```

Comment:
Non-blocking: I appreciate that it's very clear to see the entire process in this one method, but I'm curious if we'd get better naming by abstracting some of this logic into separate methods. It's very minor and I'm not sure I'd actually do it, but this method ends up doing quite a lot of very small things. I wonder if 2 or 3 methods would allow each to be named clearly and fully documented tested in a way that would be worth the extra "noise" the added complexity would introduce.

Reply:
While I appreciate the value of shorter methods, and this one is currently a bit long, I'm not sure what the natural breaking points would be. There's a "too far" region where we're defining our own methods to match those of the tools we're using, which feels like it would force us to refactor heavily if we ever change the workflow at all.

That said, one natural split point would seem to be around the tokenization step, maybe? One method to prep the submitted phrase, and then one method to combine and reorder everything?

Reply:
Another consideration - is there ever going to be a point where we'd need to call only one of these potential methods? I struggle to think of one.

Reply:
Regardless of whether we ever call these pieces independently, it is easier to _read_ the method if we create well named small methods to split up complex processes. I do agree that this is right on the edge of being complex enough to make a difference. If you use better naming within this method as suggested in other comments, I suspect my inclination to split it into smaller methods would decline.